### PR TITLE
Lookup: add lookup feature to `group_by`

### PR DESCRIPTION
**Pull request by coszio**
Adds the lookup feature to the existing `group_by` infrastructure

## Dependencies
Builds on top of #1981 

## TODOs

- [x] Create `GroupBy` builder
- [x] Add tests. Only happy paths

### All Submissions:

* [x] Have you followed the guidelines in our Contributing document?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?

### New Feature Submissions:

1. [x] Does your submission pass tests?
2. [x] Have you formatted your code locally using `cargo fmt` command prior to submission?
3. [x] Have you checked your code using `cargo clippy` command?

### Changes to Core Features:
* [x] Have you written new tests for your core changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

### CODE REVIEW

File: lib/collection/src/grouping/builder.rs [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+            // Put the lookups in their respective groups
+            for group in &mut groups {
+                if let Some(lookup) = lookups.remove(&PseudoId::from(group.id.clone())) {
+                    group.lookup = Some(lookup);
+                } else {
+                    // We want to explicitly say that there is no lookup for this group id,
+                    // this is to differentiate between not finding a lookup and not requesting a lookup
+                    group.lookup = Some(Lookup::None);
+                }
+            }
```
Comment (timvisee) [inline lib/collection/src/grouping/builder.rs:100]: Maybe we can use an iterator here:

```suggestion
            // Put the lookups in their respective groups
            groups.iter_mut().for_each(|group| {
                // If there is no lookup for this group id, we want to explicitly say that
                // (`Lookup::None`). This is to differentiate between not finding a lookup and not
                // requesting a lookup.
                group.lookup.replace(lookups.remove(&PseudoId::from(group.id.clone())).unwrap_or(Lookup::None));
            });
```

**Review by timvisee — APPROVED** [review]
Great job on the extensive testing.

I don't know all the fine details of grouping (yet!), so someone else should review this as well.

File: lib/collection/src/grouping/builder.rs (hunk 1 above)
Comment (coszio) [reply lib/collection/src/grouping/builder.rs:100]: I still struggle deciding if I like `for_each` or `for in` better. I'll try this

Comment (coszio) [reply lib/collection/src/grouping/builder.rs]: Kept it for using `Option::replace()` 👍

File: lib/collection/src/operations/types.rs [hunk 2]
```diff
@@ -409,6 +411,17 @@ pub struct RecommendGroupsRequest {
...
 
+#[derive(Debug, Serialize, Deserialize, JsonSchema, Clone)]
+pub struct PointGroup {
+    /// Scored points that have the same value of the group_by key
+    pub hits: Vec<ScoredPoint>,
+    /// Value of the group_by key, shared across all the hits in the group
+    pub id: GroupId,
+    /// Record that has been looked up using the group id
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub lookup: Option<RetrievedLookup>,
```
Comment (coszio) [inline lib/collection/src/operations/types.rs:422]: Make straight type without enum

Comment (coszio) [reply lib/collection/src/operations/types.rs]: Done

**Review by generall — APPROVED** [review]

**Pull request merged**